libraries/GCS_MAVLink/GCS.cpp

```diff
@@ -79,3 +79,72 @@ bool GCS::install_alternative_protocol(mavlink_channel_t c, GCS_MAVLINK::protoco
 }
 
 #undef FOR_EACH_ACTIVE_CHANNEL
+
+void GCS::update_sensor_status_flags()
+{
+    control_sensors_present = 0;
+    control_sensors_enabled = 0;
+    control_sensors_health = 0;
+
+    AP_AHRS &ahrs = AP::ahrs();
+    const AP_InertialSensor &ins = AP::ins();
+
+    control_sensors_present |= MAV_SYS_STATUS_AHRS;
+    control_sensors_enabled |= MAV_SYS_STATUS_AHRS;
+    if (!ahrs.initialised() || ahrs.healthy()) {
+        if (!ahrs.have_inertial_nav() || ins.accel_calibrated_ok_all()) {
+            control_sensors_health |= MAV_SYS_STATUS_AHRS;
+        }
+    }
+
+    const AP_Baro &barometer = AP::baro();
+    control_sensors_present |= MAV_SYS_STATUS_SENSOR_ABSOLUTE_PRESSURE;
+    control_sensors_enabled |= MAV_SYS_STATUS_SENSOR_ABSOLUTE_PRESSURE;
+    if (barometer.all_healthy()) {
+        control_sensors_health |= MAV_SYS_STATUS_SENSOR_ABSOLUTE_PRESSURE;
+    }
+
+    const AP_BattMonitor &battery = AP::battery();
+    control_sensors_present |= MAV_SYS_STATUS_SENSOR_BATTERY;
+    if (battery.num_instances() > 0) {
+        control_sensors_enabled |= MAV_SYS_STATUS_SENSOR_BATTERY;
+    }
+    if (battery.healthy() && !battery.has_failsafed()) {
+        control_sensors_health |= MAV_SYS_STATUS_SENSOR_BATTERY;
+    }
+
+    control_sensors_present |= MAV_SYS_STATUS_SENSOR_3D_GYRO;
+    if (!ins.calibrating()) {
+        control_sensors_enabled |= MAV_SYS_STATUS_SENSOR_3D_GYRO;
+        if (ins.get_gyro_health_all() && ins.gyro_calibrated_ok_all()) {
+            control_sensors_health |= MAV_SYS_STATUS_SENSOR_3D_GYRO;
+        }
+    }
+    control_sensors_present |= MAV_SYS_STATUS_SENSOR_3D_ACCEL;
+    if (!ins.calibrating()) {
+        control_sensors_enabled |= MAV_SYS_STATUS_SENSOR_3D_ACCEL;
+        if (ins.get_accel_health_all()) {
+            control_sensors_health |= MAV_SYS_STATUS_SENSOR_3D_ACCEL;
+        }
+    }
+
+    const AP_Logger &logger = AP::logger();
+    if (logger.logging_present()) {  // primary logging only (usually File)
+        control_sensors_present |= MAV_SYS_STATUS_LOGGING;
+    }
+    if (logger.logging_enabled()) {
```

Comment:
The downside to this approach is that if logging has failed the user doesn't get a warning until they hit the arming check (or try to log while disarmed). If present and enabled are more tightly coupled then the warning can come up earlier for users.

I'd probably complain more but I always log while disarmed so this isn't as terribly upsetting for me as it otherwise would be :)

Reply:
Problem is that we logging isn't enabled until we arm for most people.

I don't think this is a behaviour change you're point out?!

Reply:
> Problem is that we logging isn't enabled until we arm for most people.

Yeah this is sorta the classic problem: if the subsystem is ready to log once needed is this enabled or not?

Doesn't look to be a behavior change, I was mostly just pointing out a restriction to this. Since I always log while disarmed (I still hold that should be default behavior, but that's a very different conversation) I'm okay with the reporting during this stage. (Basically this comment can be dismissed/resolved, I just wanted to have the conversation about what this means/has meant from a UX perspective)